sdk/metric/pipeline_test.go

```diff
@@ -518,3 +536,91 @@ func TestExemplars(t *testing.T) {
 		check(t, r, 2, 2, 2)
 	})
 }
+
+func TestAddingAndObservingMeasureConcurrency(t *testing.T) {
+	exp := &fnExporter{}
+	r1 := NewPeriodicReader(exp, WithInterval(1*time.Second))
+	r2 := NewPeriodicReader(exp, WithInterval(600*time.Second))
+
+	mp := NewMeterProvider(WithReader(r1), WithReader(r2))
+	m := mp.Meter("test")
+
+	oc1, err := m.Int64ObservableCounter("int64-observable-counter")
+	require.NoError(t, err)
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_, err := m.Int64ObservableCounter("int64-observable-counter-2")
+		require.NoError(t, err)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_, err := m.RegisterCallback(
+			func(_ context.Context, o metric.Observer) error {
+				o.ObserveInt64(oc1, 2)
+				return nil
+			}, oc1)
+		require.NoError(t, err)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_ = mp.pipes[0].produce(context.Background(), &metricdata.ResourceMetrics{})
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_ = mp.pipes[1].produce(context.Background(), &metricdata.ResourceMetrics{})
+	}()
+
+	wg.Wait()
+}
+
+func TestPipelineWithMultipleReaders(t *testing.T) {
+	exp := &fnExporter{}
+	r1 := NewPeriodicReader(exp, WithInterval(1*time.Second))
+	r2 := NewPeriodicReader(exp, WithInterval(600*time.Second))
+
+	mp := NewMeterProvider(WithReader(r1), WithReader(r2))
+	m := mp.Meter("test")
+
+	var val atomic.Int64
+	val.Add(1)
+	measure := func(_ context.Context, m metric.Meter) {
+		oc, err := m.Int64ObservableCounter("int64-observable-counter")
+		require.NoError(t, err)
+		_, err = m.RegisterCallback(
+			// SDK periodically calls this function to collect data.
+			func(_ context.Context, o metric.Observer) error {
+				o.ObserveInt64(oc, val.Load())
+				return nil
+			}, oc)
+		require.NoError(t, err)
+	}
+	ctx := context.Background()
+	measure(ctx, m)
+	rm := new(metricdata.ResourceMetrics)
+	val.Add(1)
+
+	// adding sleep deliberately so that the callback get triggered
+	time.Sleep(2 * time.Second)
+	err := r1.Collect(ctx, rm)
+	require.NoError(t, err)
+	assert.Len(t, rm.ScopeMetrics, 1)
+	assert.Len(t, rm.ScopeMetrics[0].Metrics, 1)
+	assert.Equal(t, int64(2), rm.ScopeMetrics[0].Metrics[0].Data.(metricdata.Sum[int64]).DataPoints[0].Value)
```

Comment:
Can we use https://pkg.go.dev/github.com/stretchr/testify/assert#EventuallyWithT instead to make the test less flaky (we can put longer `waitFor`) and faster (we can put shorter `tick`).

Reply:
@pellared It should not be flaky as we are calling Collect but I think your suggestion is better as compared to what it was earlier, change is pushed. Thank you!

Reply:
>  It should not be flaky as we are calling Collect

I see. Then why not using `NewManualReader` instead of `NewPeriodicReader` to remove timing from the tests?

Reply:
Issue originally started with PeriodicReader so, I thought to use that. It doesn't look like that timing will have impact in this test case as it is predictable because of `Collect`(earlier version would have passed without any sleep also). But if you think it would be better to use `ManualReader` then I can change.

Reply:
Please change to a manual reader. Then it is clear that we do not need any time.Sleep or assert.Eventually in test

Reply:
My proposal: https://github.com/open-telemetry/opentelemetry-go/pull/5900/files#r1832185449

Reply:
@pellared Sure, making changes to adopt your proposal, it's much clearer version.

Reply:
I also noticed that usage of a period readers causes leaks in the tests as the readers are still working after the test finishes as they were never shut down.

Reply:
I have adopted ManualReader as per suggestion.